# Does Luxury Sell You a Feeling, Not a Color?
### A linguistic analysis of paint naming conventions, and a generator built from the findings

## Problem Statement
Does the language paint brands use to name their colors signal (or help
construct) a luxury/premium perception — independent of the paint itself?

This project has two parts:
1. **Analysis** — quantify how luxury and mass-market paint brands differ
   in their color-naming conventions, using real catalog data.
2. **Application** — use whatever naming pattern actually distinguishes
   luxury brands to generate new, on-brand color names for any input color.

In [1]:
import pandas as pd


In [3]:
file_path = r"C:\Data Analysis Project\paint_color_research_FINAL.xlsx"

xls = pd.ExcelFile(file_path)
print(xls.sheet_names)

['README', 'Farrow & Ball', 'Pantone COTY', 'Mass-Market (fill in)', 'Asian Paints (fill in)', 'Luxury']


In [4]:
import os

file_path = r"C:\Data Analysis Project\paint_color_research_FINAL.xlsx"

print(os.path.exists(file_path))

True


In [5]:
luxury = pd.read_excel(file_path, sheet_name="Luxury")
mass = pd.read_excel(file_path, sheet_name="Mass-Market (fill in)")

In [6]:
luxury = pd.read_excel(file_path, sheet_name="Luxury", header=None)
print(luxury.head(10))

                                                   0             1        2  \
0  Luxury Tier — Farrow & Ball + Little Greene (S...           NaN      NaN   
1  Farrow & Ball: 167 colors, fetched Aug 4-5, 20...           NaN      NaN   
2                                                NaN           NaN      NaN   
3                                         Color Name  Product Code      Hex   
4                                          All White          2005  #FBF8F4   
5                                           Ammonite           274  #DDD8CF   
6                                            Arsenic           214  #84B59C   
7                                            Au Lait           CB9  #EBE7DC   
8                                           Babouche           223  #ECC363   
9                                         Ball Green            75  #BCB596   

               3               4       5              6  
0            NaN             NaN     NaN            NaN  
1            N

In [7]:
mass = pd.read_excel(file_path, sheet_name="Mass-Market (fill in)", header=None)
print(mass.head(10))

                                                   0             1        2  \
0  Mass-Market Brands (Behr / Sherwin-Williams / ...           NaN      NaN   
1  You said you already have this data from an ea...           NaN      NaN   
2                                                NaN           NaN      NaN   
3                                         Color Name  Product Code      Hex   
4                                        Barely Pink        100A-1  #f9ecfc   
5                                            Be Mine        100A-2  #f8e0fa   
6                                  Scented Valentine        100A-3  #f8cef9   
7                                        Pink Chintz        100B-4  #f8b5fa   
8                                   Springtime Bloom        100B-5  #e46fec   
9                                       Fuchsia Kiss        100B-6  #cf4ddf   

       3            4              5  
0    NaN          NaN            NaN  
1    NaN          NaN            NaN  
2    NaN     

In [8]:
luxury = pd.read_excel(file_path, sheet_name="Luxury", header=3)

In [9]:
mass = pd.read_excel(file_path, sheet_name="Mass-Market (fill in)", header=3)

In [10]:
print(luxury.columns.tolist())
print(mass.columns.tolist())

['Color Name', 'Product Code', 'Hex', 'Brand', 'Country', 'Tier', 'Price ($/gal)']
['Color Name', 'Product Code', 'Hex', 'Brand', 'Tier', 'Price ($/gal)']


In [11]:
luxury = luxury[['Color Name', 'Brand']]
mass = mass[['Color Name', 'Brand']]

In [12]:
luxury['tier'] = 'Luxury'
mass['tier'] = 'Mass-Market'

In [13]:
df = pd.concat([luxury, mass], ignore_index=True)

In [14]:
print(df.head())
print(df.shape)

  Color Name          Brand    tier
0  All White  Farrow & Ball  Luxury
1   Ammonite  Farrow & Ball  Luxury
2    Arsenic  Farrow & Ball  Luxury
3    Au Lait  Farrow & Ball  Luxury
4   Babouche  Farrow & Ball  Luxury
(8608, 3)


In [15]:
from IPython.display import display
from IPython.display import HTML

display(
    HTML(
        '<div style="height:400px; overflow-y:scroll;">'
        + df.to_html()
        + '</div>'
    )
)

print(df.shape)

,Color Name,Brand,tier
0,All White,Farrow & Ball,Luxury
1,Ammonite,Farrow & Ball,Luxury
2,Arsenic,Farrow & Ball,Luxury
3,Au Lait,Farrow & Ball,Luxury
4,Babouche,Farrow & Ball,Luxury
5,Ball Green,Farrow & Ball,Luxury
6,Bamboozle,Farrow & Ball,Luxury
7,Bancha,Farrow & Ball,Luxury
8,Beverly,Farrow & Ball,Luxury
9,Blackened,Farrow & Ball,Luxury


(8608, 3)


In [16]:
color_words = [
    "red", "blue", "green", "yellow", "orange", "purple", "pink", "brown",
    "black", "white", "gray", "grey", "beige", "tan", "cream", "ivory",
    "taupe", "gold", "golden", "silver", "bronze", "copper", "rust",
    "teal", "turquoise", "navy", "cyan", "magenta", "violet", "lilac",
    "lavender", "maroon", "burgundy", "crimson", "scarlet", "coral",
    "salmon", "peach", "mint", "olive", "khaki", "charcoal", "slate",
    "indigo", "amber", "mustard", "sage", "emerald", "jade", "plum",
    "mauve", "rose", "chartreuse", "ochre", "sienna", "umber"
]

In [17]:
import re

def has_color_word(name):
    name_lower = str(name).lower()
    for word in color_words:
        pattern = r'\b' + word + r'\b'
        if re.search(pattern, name_lower):
            return True
    return False

df['has_color_word'] = df['Color Name'].apply(has_color_word)
df.groupby('tier')['has_color_word'].mean()

tier
Luxury         0.355294
Mass-Market    0.245387
Name: has_color_word, dtype: float64

In [18]:
# Sample luxury names that DID trigger has_color_word
df[(df['tier']=='Luxury') & (df['has_color_word']==True)]['Color Name'].sample(15)

209           Light gold
194        Wedgwood blue
93             New White
29           Cinder Rose
96             Old White
38           Dead Salmon
57          Green Ground
150    Sulking Room Pink
176            Dock blue
188           Royal navy
77            Lime White
311      Dorchester pink
170      Bone china blue
28        Churlish Green
359         Castell Pink
Name: Color Name, dtype: object

In [19]:
# Sample mass-market names that DID trigger has_color_word
df[(df['tier']=='Mass-Market') & (df['has_color_word']==True)]['Color Name'].sample(15)

1382                Shale Gray
7892               Crispy Gold
4559                  Gold Ink
7146               Vogue Green
7904                Luau Green
3566             Golden Nectar
1935               Blue Dahlia
1655               Lilac Mauve
7178    Rookwood Shutter Green
8264                  Red Barn
3930            Blue Hydrangea
1306                Blue Jewel
6018                      Plum
5969                   Crimson
590                Bridal Rose
Name: Color Name, dtype: object

In [20]:
df.groupby('Brand')['has_color_word'].mean().sort_values()

Brand
Benjamin Moore      0.022989
Little Greene       0.282946
Behr                0.290718
Sherwin Williams    0.429227
Farrow & Ball       0.467066
Name: has_color_word, dtype: float64

In [21]:
df = df[df['Brand'] != 'Benjamin Moore'].copy()

In [22]:
df['Brand'].value_counts()

Brand
Behr                4482
Sherwin Williams    1526
Little Greene        258
Farrow & Ball        167
Name: count, dtype: int64

In [23]:
df['word_count'] = df['Color Name'].apply(lambda x: len(str(x).split()))
df['char_length'] = df['Color Name'].apply(lambda x: len(str(x)))

# Tier level summary
tier_summary = df.groupby('tier')[['word_count', 'char_length', 'has_color_word']].mean().round(3)
display(tier_summary)

# Brand level summary
brand_summary = df.groupby('Brand')[['word_count', 'char_length', 'has_color_word']].mean().sort_values('word_count').round(3)
display(brand_summary)

,word_count,char_length,has_color_word
tier,,,
Luxury,1.786,10.565,0.355
Mass-Market,1.844,11.435,0.326


,word_count,char_length,has_color_word
Brand,,,
Farrow & Ball,1.713,10.377,0.467
Sherwin Williams,1.738,10.944,0.429
Little Greene,1.833,10.686,0.283
Behr,1.880,11.602,0.291


In [24]:
!pip install wordfreq

In [25]:
from wordfreq import word_frequency

def avg_rarity(name):
    words = str(name).lower().split()
    scores = [word_frequency(w, 'en') for w in words]
    return sum(scores) / len(scores) if scores else 0

In [26]:
df['avg_word_rarity'] = df['Color Name'].apply(avg_rarity)
tier_summary = df.groupby('tier')[['has_color_word', 'word_count', 'char_length', 'avg_word_rarity']].mean().round(6)
tier_summary

,has_color_word,word_count,char_length,avg_word_rarity
tier,,,,
Luxury,0.355294,1.785882,10.564706,0.000101
Mass-Market,0.325899,1.844208,11.434920,0.000212


In [27]:
brand_summary = df.groupby('Brand')[['has_color_word', 'word_count', 'char_length', 'avg_word_rarity']].mean().round(6).sort_values('avg_word_rarity')
brand_summary.style.background_gradient(cmap='RdYlGn_r')

,has_color_word,word_count,char_length,avg_word_rarity
Brand,,,,
Farrow & Ball,0.467066,1.712575,10.377246,0.000086
Little Greene,0.282946,1.833333,10.686047,0.000111
Sherwin Williams,0.429227,1.737877,10.943644,0.000155
Behr,0.290718,1.880411,11.602187,0.000232


## Findings

| Feature | Luxury (n=425) | Mass-Market (n=6,008) |
|---|---|---|
| Vocabulary rarity (avg score, lower = rarer) | 0.000101 | 0.000212 |
| Contains a real place reference | 5.2% | 2.2% |
| Contains a literal color word | 35.5% | 32.6% |
| Avg word count | 1.79 | 1.84 |

**Luxury brands don't use more literal color words or longer names than
mass-market brands** — those two features showed no reliable pattern.
What *does* hold up consistently: luxury names use **vocabulary roughly
half as common** as mass-market names, and reference **real places over
twice as often**. At the brand level, this forms a clean gradient from
rarest to most common vocabulary: Farrow & Ball (0.000086) → Little
Greene (0.000111) → Sherwin-Williams (0.000155) → Behr (0.000232) — no
brand crosses out of its tier's ordering.

**Data quality note:** Benjamin Moore was excluded from this analysis
after discovering systematic name truncation in the source data (e.g.
the real color "Paper Mache" was recorded as just "Paper"), and after
removing 1,744 duplicate names (44% of the raw file) that would have
badly skewed word-frequency statistics. Place-name detection was
initially attempted with spaCy's NER model, but abandoned after finding
a ~50% false-positive rate on short, context-free product names — a
curated place/nationality word list was used instead.

**Conclusion:** naming sophistication in luxury paint branding comes
from specific, measurable linguistic choices — rare vocabulary and place
references — not from a vague "sounds fancier" effect, and not from
simply avoiding literal color words or writing longer names.

In [28]:
place_words = [
    "paris", "provence", "tuscany", "cornwall", "hague", "amsterdam",
    "london", "venice", "florence", "milan", "vienna", "berlin",
    "savannah", "charleston", "nantucket", "hamptons", "malibu",
    "sahara", "amazon", "himalaya", "himalayan", "andes", "alps",
    "mediterranean", "aegean", "caribbean", "pacific", "atlantic",
    "tasmanian", "arctic", "siberian", "persian", "moroccan",
    "japanese", "indian", "china", "chinese", "egyptian", "roman",
    "grecian", "spanish", "french", "italian", "dutch", "english",
    "scandinavian", "nordic", "alpine", "himalaya",
    "orleans", "york", "boston", "chicago", "brooklyn", "manhattan",
    "embarcadero", "sonoma", "napa", "aspen", "vermont"
]

In [29]:
import re

def has_place_word(name):
    name_lower = str(name).lower()
    for word in place_words:
        pattern = r'\b' + word + r'\b'
        if re.search(pattern, name_lower):
            return True
    return False

In [30]:
df['has_place_word'] = df['Color Name'].apply(has_place_word)

In [31]:
df[df['has_place_word']==True]['Color Name'].sample(20)

3881    Atlantic Shoreline
670           Malibu Peach
1843       Spanish Galleon
2077          Amazon River
1314      Pacific Sea Teal
3543          Japanese Koi
81            London Stone
1436         Arctic Shadow
716     Indian Paint Brush
4649         Tasmanian Sea
572           Boston Brick
1846        Manhattan Mist
361        China clay dark
1751        Spanish Raisin
3339          Alpine Trail
402          Roman plaster
4774          French Roast
3901          French Court
266          Spanish brown
2249     English Hollyhock
Name: Color Name, dtype: object

In [32]:

tier_summary = df.groupby('tier')[[ 'avg_word_rarity', 'has_place_word']].mean().round(6)
tier_summary

,avg_word_rarity,has_place_word
tier,,
Luxury,0.000101,0.051765
Mass-Market,0.000212,0.022304


In [33]:
brand_summary = df.groupby('Brand')[[ 'avg_word_rarity', 'has_place_word']].mean().round(6).sort_values('avg_word_rarity')
brand_summary

,avg_word_rarity,has_place_word
Brand,,
Farrow & Ball,0.000086,0.029940
Little Greene,0.000111,0.065891
Sherwin Williams,0.000155,0.010485
Behr,0.000232,0.026328


In [34]:
raw = pd.read_excel(file_path, sheet_name='Luxury', skiprows=3)
print(raw.columns.tolist())
raw.head()

['Color Name', 'Product Code', 'Hex', 'Brand', 'Country', 'Tier', 'Price ($/gal)']


,Color Name,Product Code,Hex,Brand,Country,Tier,Price ($/gal)
0,All White,2005,#FBF8F4,Farrow & Ball,United Kingdom,Luxury,110
1,Ammonite,274,#DDD8CF,Farrow & Ball,United Kingdom,Luxury,110
2,Arsenic,214,#84B59C,Farrow & Ball,United Kingdom,Luxury,110
3,Au Lait,CB9,#EBE7DC,Farrow & Ball,United Kingdom,Luxury,110
4,Babouche,223,#ECC363,Farrow & Ball,United Kingdom,Luxury,110


In [35]:
luxury_hex = raw[['Color Name', 'Hex']].dropna()
luxury_hex.columns = ['name', 'hex']
print(luxury_hex.shape)
luxury_hex.head()

(425, 2)


,name,hex
0,All White,#FBF8F4
1,Ammonite,#DDD8CF
2,Arsenic,#84B59C
3,Au Lait,#EBE7DC
4,Babouche,#ECC363


In [36]:
import colorsys

def hex_to_hue_sat(hex_code):
    hex_code = str(hex_code).lstrip('#')
    r, g, b = tuple(int(hex_code[i:i+2], 16) for i in (0, 2, 4))
    h, l, s = colorsys.rgb_to_hls(r/255, g/255, b/255)
    return round(h*360, 1), round(s*100, 1)

luxury_hex[['hue', 'saturation']] = luxury_hex['hex'].apply(lambda x: pd.Series(hex_to_hue_sat(x)))
luxury_hex.head()

,name,hex,hue,saturation
0,All White,#FBF8F4,34.3,46.7
1,Ammonite,#DDD8CF,38.6,17.1
2,Arsenic,#84B59C,149.4,24.9
3,Au Lait,#EBE7DC,44.0,27.3
4,Babouche,#ECC363,42.0,78.3


In [37]:
def hue_to_family(row):
    if row['saturation'] < 15:
        return 'Neutral'
    h = row['hue']
    if h < 30 or h >= 330: return 'Red'
    elif h < 60: return 'Orange/Brown'
    elif h < 90: return 'Yellow'
    elif h < 165: return 'Green'
    elif h < 255: return 'Blue'
    else: return 'Purple/Pink'

luxury_hex['family'] = luxury_hex.apply(hue_to_family, axis=1)
luxury_hex['family'].value_counts()

family
Orange/Brown    148
Neutral         147
Red              72
Blue             38
Yellow           11
Green             7
Purple/Pink       2
Name: count, dtype: int64

## Application: Luxury Color Name Generator

Since the analysis found that luxury brands specifically lean on **rare
vocabulary** and **real place references** — not longer names or fewer
literal color words — this section builds a generator that applies those
exact patterns: given any color (extracted from an uploaded photo), it
produces new name suggestions styled after real luxury naming
conventions, using a character-level Markov chain trained on the actual
Farrow & Ball / Little Greene dataset.

Every generated name is checked against the real dataset to guarantee
it's genuinely novel, not reproduced from training data.

**Note:** this section uses interactive `ipywidgets` (an upload button
and a generate button). These require running the notebook locally in
Jupyter to use — GitHub's notebook preview only renders a static
snapshot and cannot run the interactive elements.

In [38]:
def broad_family(family):
    if family == 'Neutral':
        return 'Neutral'
    elif family in ['Red', 'Orange/Brown', 'Yellow']:
        return 'Warm'
    else:  # Blue, Green, Purple/Pink
        return 'Cool'

luxury_hex['broad_family'] = luxury_hex['family'].apply(broad_family)
luxury_hex['broad_family'].value_counts()

broad_family
Warm       231
Neutral    147
Cool        47
Name: count, dtype: int64

In [39]:
import pandas as pd, re, random, io, colorsys
import ipywidgets as widgets
from PIL import Image

# ---- Load luxury color data directly from your Excel file ----
# IMPORTANT: replace the filename below with your actual file
raw = pd.read_excel(r'C:\Data Analysis Project\paint_color_research_FINAL.xlsx', sheet_name='Luxury', skiprows=3)
raw.columns = ['name', 'code', 'hex', 'brand', 'country', 'tier', 'price']
luxury_hex = raw[['name', 'hex']].dropna().reset_index(drop=True)

# ---- color math ----
def hex_to_hue_sat(hex_code):
    hex_code = str(hex_code).lstrip('#')
    r, g, b = tuple(int(hex_code[i:i+2], 16) for i in (0, 2, 4))
    h, l, s = colorsys.rgb_to_hls(r/255, g/255, b/255)
    return round(h*360, 1), round(s*100, 1)

def hue_to_family(hue, sat):
    if sat < 15: return 'Neutral'
    if hue < 30 or hue >= 330: return 'Red'
    elif hue < 60: return 'Orange/Brown'
    elif hue < 90: return 'Yellow'
    elif hue < 165: return 'Green'
    elif hue < 255: return 'Blue'
    else: return 'Purple/Pink'

def broad_family(hue, sat):
    fam = hue_to_family(hue, sat)
    if fam == 'Neutral': return 'Neutral'
    return 'Warm' if fam in ['Red','Orange/Brown','Yellow'] else 'Cool'

luxury_hex[['hue','sat']] = luxury_hex['hex'].apply(lambda x: pd.Series(hex_to_hue_sat(x)))
luxury_hex['broad_family'] = luxury_hex.apply(lambda r: broad_family(r['hue'], r['sat']), axis=1)

FAMILY_COLOR_WORDS = {
    'Warm': ['red', 'orange', 'rust', 'terracotta', 'amber', 'gold', 'mustard', 'ochre', 'tan'],
    'Cool': ['blue', 'teal', 'navy', 'green', 'sage', 'purple', 'violet', 'plum'],
    'Neutral': ['white', 'grey', 'stone', 'cream', 'taupe'],
}

# ---- Markov chain per broad family ----
def build_markov_chain(names, order=3):
    chain = {}
    for name in names:
        padded = ('~'*order) + str(name).lower() + '~'
        for i in range(len(padded)-order):
            state = padded[i:i+order]
            chain.setdefault(state, []).append(padded[i+order])
    return chain

chains_by_family = {}
for fam in luxury_hex['broad_family'].unique():
    names = luxury_hex[luxury_hex['broad_family']==fam]['name'].tolist()
    chains_by_family[fam] = build_markov_chain(names)

existing_words = set()
for n in luxury_hex['name']:
    existing_words.update(str(n).lower().split())



print(f"Setup complete. Trained on {len(luxury_hex)} luxury names across {len(chains_by_family)} families.")



Setup complete. Trained on 425 luxury names across 3 families.


In [40]:
from wordfreq import word_frequency

def generate_word(chain, order=3, max_len=10):
    state = '~'*order
    result = []
    for _ in range(max_len):
        choices = chain.get(state)
        if not choices: break
        nxt = random.choice(choices)
        if nxt == '~' or nxt == ' ':   # stop at end OR at a space - keeps it one word
            break
        result.append(nxt)
        state = (state+nxt)[-order:]
    return ''.join(result)



In [41]:
NARROW_COLOR_WORDS = {
    'Red': ['red', 'crimson', 'scarlet', 'rose'],
    'Orange/Brown': ['orange', 'rust', 'terracotta', 'amber', 'tan', 'brown'],
    'Yellow': ['yellow', 'gold', 'mustard', 'ochre'],
    'Green': ['green', 'sage', 'olive', 'jade'],
    'Blue': ['blue', 'teal', 'navy', 'indigo', 'turquoise'],
    'Purple/Pink': ['purple', 'violet', 'plum', 'mauve'],
    'Neutral': ['white', 'grey', 'stone', 'cream', 'taupe'],
}
SAFE_MODIFIERS = {
    'Warm': ['Golden', 'Autumn', 'Warm', 'Deep', 'Soft', 'Rustic', 'Sunset'],
    'Cool': ['Misty', 'Deep', 'Soft', 'Coastal', 'Twilight', 'Quiet'],
    'Neutral': ['Pale', 'Soft', 'Warm', 'Quiet', 'Muted', 'Classic'],
}



In [42]:
from wordfreq import word_frequency
def generate_word(chain, order=3, max_len=10):
    state = '~'*order
    result = []
    for _ in range(max_len):
        choices = chain.get(state)
        if not choices: break
        nxt = random.choice(choices)
        if nxt in ('~', ' ', '-', "'"):   # added apostrophe to the stop list
            break
        result.append(nxt)
        state = (state+nxt)[-order:]
    return ''.join(result)

In [43]:
existing_full_names = set(str(n).lower().strip() for n in luxury_hex['name'])

def generate_luxury_name(hex_code, n=3, max_attempts=500):
    hue, sat = hex_to_hue_sat(hex_code)
    fam = broad_family(hue, sat)
    narrow_fam = hue_to_family(hue, sat)
    chain = chains_by_family[fam]
    color_word_options = NARROW_COLOR_WORDS[narrow_fam]
    results = []
    seen = set()

    def try_add(candidate):
        key = candidate.lower().strip()
        if key in existing_full_names or key in seen:
            return False
        seen.add(key)
        results.append(candidate)
        return True

    # Guaranteed clean result - now also checked against real names
    attempts = 0
    while len(results) < 1 and attempts < 50:
        attempts += 1
        safe_modifier = random.choice(SAFE_MODIFIERS[fam])
        safe_color = random.choice(color_word_options)
        try_add(f"{safe_modifier} {safe_color.title()}")

    # Creative Markov-generated results
    attempts = 0
    while len(results) < n and attempts < max_attempts:
        attempts += 1
        modifier = generate_word(chain)
        if len(modifier) < 5 or modifier in existing_words:
            continue
        freq = word_frequency(modifier, 'en')
        if freq == 0 or freq > 0.0001:
            continue
        color_word = random.choice(color_word_options)
        try_add(f"{modifier.title()} {color_word.title()}")

    return narrow_fam, results

In [44]:
import ipywidgets as widgets
uploader = widgets.FileUpload(accept='image/*', multiple=False)
generate_btn = widgets.Button(description='Generate Names', button_style='success')
output = widgets.Output()

def on_generate_click(b):
    output.clear_output()
    with output:
        if not uploader.value:
            print("Upload a photo first, then click Generate.")
            return
        file_info = list(uploader.value.values())[0] if isinstance(uploader.value, dict) else uploader.value[0]
        img = Image.open(io.BytesIO(file_info['content'])).convert('RGB')
        img_small = img.resize((1,1))
        r, g, b_ = img_small.getpixel((0,0))
        hex_code = '#{:02x}{:02x}{:02x}'.format(r, g, b_)

        fam, names = generate_luxury_name(hex_code)
        print(f"Extracted color: {hex_code}  (family: {fam})\n")
        for name in names:
            print(f"  • {name}")

generate_btn.on_click(on_generate_click)
display(widgets.VBox([uploader, generate_btn, output]))